### PR TITLE
HPCC-19544 Clean up CKeyCursor ahead of migration to new field filters

### PR DESCRIPTION
**Pull request by richardkchapman**
Also remove some unused code, reduce memory footprint of CKeyCursor objects,
and generally clean things up a bit.

Also addresses HPCC-19545 Remove overrideable segmonitor support

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Roxie regression suite clean.

### CODE REVIEW

File: system/jhtree/jhtree.ipp [hunk 1]
```diff
@@ -167,37 +167,76 @@ public:
 
 class jhtree_decl CKeyCursor : public IKeyCursor, public CInterface
```
Comment (ghalliday) [inline system/jhtree/jhtree.ipp]: minor: CInterfaceOf<IKeyCursor> would save 8 bytes per cursor.

File: system/jhtree/jhtree.cpp [hunk 2]
```diff
@@ -1949,56 +1658,304 @@ bool CKeyCursor::ltEqual(const char *src, char *dst, bool seekForward)
...
+            unsigned __int64 locount = getSequence();
+            endRange(lastFullSeg);
+            ltEqual(keyBuffer, stats);
+            lseeks++;
+            result += getSequence()-locount+1;
+            if (max && (result > max))
+                break;
+            if (!incrementKey(lastFullSeg))
+                break;
+        }
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: This used to set matched to false - is the removal deliberate?

File: system/jhtree/jhtree.cpp [hunk 3]
```diff
@@ -1949,56 +1658,304 @@ bool CKeyCursor::ltEqual(const char *src, char *dst, bool seekForward)
...
+        if (_lookup(true, lastRealSeg, stats))
+        {
+            unsigned __int64 locount = getSequence();
+            endRange(lastRealSeg);
+            ltEqual(keyBuffer, stats);
+            lseeks++;
+            result += getSequence()-locount+1;
+            if (!incrementKey(lastRealSeg))
+                break;
+        }
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: this also used to set matched to false.  Is that still needed?

File: system/jhtree/jhtree.hpp [hunk 4]
```diff
@@ -38,22 +38,49 @@ interface jhtree_decl IDelayedFile : public IInterface
...
 
+class KeyStatsCollector
+{
+public:
+    IContextLogger *ctx;
+    unsigned seeks = 0;
+    unsigned scans = 0;
+    unsigned wildseeks = 0;
+    unsigned skips = 0;
+    unsigned nullskips = 0;
```
Comment (ghalliday) [inline system/jhtree/jhtree.hpp]: trivial: could delete since no longer returned.

**Comment by ghalliday** [comment]
@richardkchapman a few minor comments

File: system/jhtree/jhtree.cpp (hunk 3 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: I don't believe so

Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: ltEqual clears it

File: system/jhtree/jhtree.cpp (hunk 2 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: ltEqual clears it. Whether that is cleaner or not, I don't know!

File: system/jhtree/jhtree.hpp (hunk 4 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.hpp]: It's returned via the log context still

File: system/jhtree/jhtree.ipp (hunk 1 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.ipp]: Will address in a future PR

**Pull request merged**